File: packages/silo-client/src/media/media-reference.ts
/**
 * The `silo://media/<id>` scheme an entry field holds to name a catalogued
 * asset — the client-side counterpart to `@silo/shared`'s `MediaRef`, not
 * imported here since this package ships zero runtime dependencies.
 *
 * Recognises only the canonical scheme form, deliberately narrower than the
 * server's `MediaRef`: a client either builds a reference itself or reads one
 * back verbatim, so accepting `asset.url` would let "I stored the wrong
 * property" keep working until the first rename.
 */
export class MediaReference {
  private static readonly Scheme = "silo://media/";

  /** Builds the reference to store in an entry field: `silo://media/<id>`. */
  static of(id: string): string {
    return MediaReference.Scheme + id;
  }

  /** The asset id a `silo://media/` reference names, or `null` if `value`
   * is not one. A trailing fragment, query or path segment is ignored. */
  static idOf(value: unknown): string | null {
    if (typeof value !== "string" || !value.startsWith(MediaReference.Scheme)) return null;
    const id = value.slice(MediaReference.Scheme.length).split(/[#?/]/)[0];
    return id ? id : null;
  }
}

File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/MediaReference.java
package in.org.quicko.silo.client.media;

/**
 * The {@code silo://media/<id>} scheme an entry field holds to name a catalogued
 * asset.
 *
 * <p>Recognises only the canonical scheme form, deliberately narrower than the
 * server's: a client either builds a reference itself or reads one back
 * verbatim, so accepting an asset's {@code url} here would let "I stored the
 * wrong property" keep working until the first rename.
 */
public final class MediaReference {
  private static final String Scheme = "silo://media/";

  private MediaReference() {}

  /** The reference to store in an entry field. */
  public static String of(String id) {
    return Scheme + id;
  }

  /** The asset id a reference names, or null when {@code value} is not one. A
   *  trailing fragment, query or path segment is ignored. */
  public static String idOf(Object value) {
    if (!(value instanceof String text) || !text.startsWith(Scheme)) return null;
    String rest = text.substring(Scheme.length());
    int end = rest.length();
    for (int index = 0; index < rest.length(); index++) {
      char character = rest.charAt(index);
      if (character == '#' || character == '?' || character == '/') {
        end = index;
        break;
      }
    }
    String id = rest.substring(0, end);
    return id.isEmpty() ? null : id;
  }
}
